### ISSUE TITLE
NPE when joining the cluster

### ISSUE THREAD

**Issue by cstamas** [hazelcast/hazelcast#8733]
I had a node (call it `nodeB`) that was joining to cluster, consisting of one another node, call it `nodeA`, so cluster initially had 1 member. During join, I had this exception logged:

```
jvm 1    | 2016-08-17 11:58:45,643+0200 INFO  [hz._hzInstance_1_hzgroup.IO.thread-Acceptor] *SYSTEM com.hazelcast.nio.tcp.SocketAcceptorThread - [127.0.0.1]:5701 [hzgroup] [3.6.4] Accepting socket connection from /127.0.0.1:51890
jvm 1    | 2016-08-17 11:58:45,643+0200 INFO  [cached7] *SYSTEM com.hazelcast.nio.tcp.TcpIpConnectionManager - [127.0.0.1]:5701 [hzgroup] [3.6.4] Established socket connection between /127.0.0.1:5701 and /127.0.0.1:51890
jvm 1    | 2016-08-17 11:58:47,995+0200 INFO  [hz._hzInstance_1_hzgroup.migration] *SYSTEM com.hazelcast.partition.InternalPartitionService - [127.0.0.1]:5701 [hzgroup] [3.6.4] Partition balance is ok, no need to re-partition cluster data... 
jvm 1    | 2016-08-17 11:58:51,655+0200 INFO  [hz._hzInstance_1_hzgroup.generic-operation.thread-2] *SYSTEM com.hazelcast.cluster.ClusterService - [127.0.0.1]:5701 [hzgroup] [3.6.4] 
jvm 1    | 
jvm 1    | Members [2] {
jvm 1    | 	Member [127.0.0.1]:5701 this
jvm 1    | 	Member [127.0.0.1]:5702
jvm 1    | }
jvm 1    | 
jvm 1    | 2016-08-17 11:58:51,765+0200 INFO  [hz._hzInstance_1_hzgroup.migration] *SYSTEM com.hazelcast.partition.InternalPartitionService - [127.0.0.1]:5701 [hzgroup] [3.6.4] Re-partitioning cluster data... Migration queue size: 135
jvm 1    | 2016-08-17 11:58:52,051+0200 WARN  [hz._hzInstance_1_hzgroup.partition-operation.thread-7] *SYSTEM com.hazelcast.partition.impl.ReplicaSyncRequest - [127.0.0.1]:5701 [hzgroup] [3.6.4] null
jvm 1    | java.lang.NullPointerException: null
jvm 1    | 	at com.hazelcast.concurrent.atomiclong.AtomicLongService.prepareReplicationOperation(AtomicLongService.java:106) ~[hazelcast-3.6.4.jar:3.6.4]
jvm 1    | 	at com.hazelcast.partition.impl.ReplicaSyncRequest.createReplicationOperations(ReplicaSyncRequest.java:150) ~[hazelcast-3.6.4.jar:3.6.4]
jvm 1    | 	at com.hazelcast.partition.impl.ReplicaSyncRequest.run(ReplicaSyncRequest.java:82) ~[hazelcast-3.6.4.jar:3.6.4]
jvm 1    | 	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:172) [hazelcast-3.6.4.jar:3.6.4]
jvm 1    | 	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:393) [hazelcast-3.6.4.jar:3.6.4]
jvm 1    | 	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.processPacket(OperationThread.java:184) [hazelcast-3.6.4.jar:3.6.4]
jvm 1    | 	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.process(OperationThread.java:137) [hazelcast-3.6.4.jar:3.6.4]
jvm 1    | 	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.doRun(OperationThread.java:124) [hazelcast-3.6.4.jar:3.6.4]
jvm 1    | 	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.run(OperationThread.java:99) [hazelcast-3.6.4.jar:3.6.4]
jvm 1    | 2016-08-17 11:58:53,094+0200 INFO  [hz._hzInstance_1_hzgroup.migration] *SYSTEM com.hazelcast.partition.InternalPartitionService - [127.0.0.1]:5701 [hzgroup] [3.6.4] All migration tasks have been completed, queues are empty.
```

My assumption is that this was an IAtomicLong, that `nodeA` removed (invoked `IAtomicLong.destroy()`) during replication, while `nodeB` was joining.

Hazelcast version: 3.6.4
Java: Oracle Java8
OS: dev OSX, prod Linux

**Comment by jerrinot** [comment]
hi @cstamas,

this looks like a bug. I briefly checked the code and it appears there is a race between `destroy()` and join. Just to double-check: are you dynamically creating `IAtomingLong` instances and destroying them at some point?

**Comment by cstamas** [comment]
@jerrinot Yes, that is the case: a node is dynamically creating IAtomicLongs, uses them, and using "reference counting" destroys them as well at some point.

**Comment by cstamas** [comment]
@jerrinot btw, we do the same (dynamically adding and removing at some point) with ISemaphore instances too, so maybe worth check that code too?

**Comment by jerrinot** [comment]
The good news is the NPE in `IAtomicLong`  was already fixed in 3.7: https://github.com/jerrinot/hazelcast/commit/62e11ceb5fa498d64c50ef5ff11320ea01a52c45#diff-743aa0436dd63bb95c956e8230f159f3R103

I'm going to check other structures. 

3.7 is going to be released really soon. like today ;-) (ops, here, I said it!:)

**Comment by cstamas** [comment]
woohoo! Thanks, will give it a shot, I see that 3.7-EA is already there, will give it a shot.

**Comment by jerrinot** [comment]
@cstamas: 3.7 was just released. It should be in Maven Central in 1-2 hours. (there is always some delay)

**Comment by cstamas** [comment]
@jerrinot is already in there (but you must to trick fastly, as directory listing will not show `3.7` for 1-2h, but things are _there in few minutes_) ;)

http://repo1.maven.org/maven2/com/hazelcast/hazelcast/3.7/

**Comment by jerrinot** [comment]
@cstamas: today I learned! thanks!

**Comment by jerrinot** [comment]
this is fixed in 3.7 by jerrinot@62e11ce 
Also it seems @Donnerbart did a good job here and fixed it in other parts too.
